File: go/0011.container-with-most-water/solution.go
// Created by Noirkelo at 2025/08/05 13:29
// leetgo: dev
// https://leetcode.cn/problems/container-with-most-water/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func maxArea(height []int) (ans int) {
	l := 0
	r := len(height) - 1
	for l < r {
		if height[l] < height[r] {
			current_area := height[l] * (r - l)
			ans = max(current_area, ans)
			l++
		} else {
			current_area := height[r] * (r - l)
			ans = max(current_area, ans)
			r--
		}
	}
	return
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	height := Deserialize[[]int](ReadLine(stdin))
	ans := maxArea(height)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0011.container-with-most-water/solution.rs
// Created by Noirkelo at 2025/08/05 13:30
// leetgo: dev
// https://leetcode.cn/problems/container-with-most-water/


use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::cmp::max;
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut l=0;
		let mut r=height.len()-1;
		let mut max_area=0;

		while l<r{
			if height[l]<height[r]{
				let cur_area=height[l]*(r-l)as i32;
				max_area=max(max_area,cur_area);
				l+=1;

			}else{
				let cur_area=height[r]*(r-l)as i32;
				max_area=max(max_area,cur_area);
				r-=1;
			}
			
		}
		return max_area;
    }
}

// @lc code=end

fn main() -> Result<()> {
	let height: Vec<i32> = deserialize(&read_line()?)?;
	let ans: i32 = Solution::max_area(height).into();

	println!("\noutput: {}", serialize(ans)?);
	Ok(())
}
